File: Python/0047-m-permutations-ii.py
"""
DFS and using a set in each pos to remove duplicates
Time: O(n!)
Space: O(n * 2)
"""
class Solution_1:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return []
        res = []
        visited = [False] * len(nums)
        self.helper(sorted(nums), [], visited, res)
        return res

    def helper(self, nums, path, visited, res):
        if len(path) == len(nums):
            res.append(path[:])
            return

        for i in range(len(nums)):
            if visited[i] or \
                (i != 0 and nums[i] == nums[i - 1] and not visited[i - 1]):
                continue
            visited[i] = True
            path.append(nums[i])
            self.helper(nums, path, visited, res)
            path.pop()
            visited[i] = 0


class Solution_2:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return []
        res = []
        self.helper(nums, [], 0, res)
        return res

    def helper(self, nums, lst, pos, res):
        # base case:
        if pos == len(nums):
            res.append(nums[:])
            return

        # main cases:
        for i in range(pos, len(nums)):
            # swap: fix a position for going down
            nums[pos], nums[i] = nums[i], nums[pos]
            self.helper(nums, lst, pos + 1, res)
            # swap back: free a position for going right
            nums[pos], nums[i] = nums[i], nums[pos]

File: leetcode/C++/0047-m-permutations-ii.cpp
class Solution {
public:
       
    void helper(vector<int>& nums, vector<int>& path, vector<bool>& visited, vector<vector<int>>& res) {
        if (path.size() == nums.size()) {
            res.push_back(path);
            return;
        }
        for (int i = 0; i < nums.size(); i++) {
            if (visited[i] || 
                (i != 0 && nums[i] == nums[i-1] && !visited[i-1]))
                continue;
            visited[i] = true;
            path.push_back(nums[i]);
            helper(nums, path, visited, res);
            path.pop_back();
            visited[i] = false;
        }
    }
    
    
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> res;
        if (nums.size() == 0)
            return res;
        vector<int> path;
        vector<bool> visited(nums.size(), false);
        sort(nums.begin(), nums.end());
        helper(nums, path, visited, res);
        return res;
    }
};


// DFS + Swap
class Solution {
public:
    void swap(vector<int>& nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
    
    void helper(vector<int>& nums, int pos, vector<vector<int>>& res) {
        if (pos == nums.size()) {
            res.push_back(nums);
            return;
        }
        unordered_set<int> used;
        for (int i = pos; i < nums.size(); i++) {
            if (used.find(nums[i]) == used.end()) {
                used.insert(nums[i]);
                swap(nums, i, pos);
                helper(nums, pos + 1, res);
                swap(nums, i, pos);
            }
        }
    }
    
    
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> res;
        if (nums.size() == 0)
            return res;
        helper(nums, 0, res);
        return res;
    }
};